apps/prairielearn/src/lib/groups.ts

```diff
@@ -191,88 +200,196 @@ export async function getQuestionGroupPermissions(
-    flash('error', 'You are already in another group.');
-    return;
-  }
+  const group_name = splitJoinCode[0];
+  const join_code = splitJoinCode[1].toUpperCase();
 
   try {
-    await sqldb.callAsync('group_users_insert', [
-      assessmentId,
-      userId,
-      authnUserId,
-      groupName,
-      joinCode,
-    ]);
+    await sqldb.runInTransactionAsync(async () => {
+      const group = await sqldb.queryOptionalRow(
+        sql.select_and_lock_group_by_name,
+        { group_name, assessment_id },
+        GroupSchema,
+      );
+      if (group == null || group.join_code !== join_code) {
+        throw new GroupOperationError('Group does not exist.');
+      }
+      await addUserToGroup({
+        assessment_id,
+        group_id: group.id,
+        uid,
+        authn_user_id,
+        enforceGroupSize: true,
+      });
+    });
   } catch (err) {
-    flash(
-      'error',
-      `Failed to join the group with join code ${fullJoinCode}. It is already full or does not exist. Please try to join another one.`,
-    );
+    if (err instanceof GroupOperationError) {
+      throw new GroupOperationError(`Cannot join group "${fullJoinCode}": ${err.message}`);
+    }
+    throw err;
   }
 }
 
 export async function createGroup(
-  groupName: string,
-  assessmentId: string,
-  userId: string,
-  authnUserId: string,
+  group_name: string,
+  assessment_id: string,
+  uids: string[],
+  authn_user_id: string,
 ): Promise<void> {
-  if (groupName.length > 30) {
-    flash('error', 'The group name is too long. Use at most 30 alphanumerical characters.');
-    return;
+  if (group_name.length > 30) {
+    throw new GroupOperationError(
+      'The group name is too long. Use at most 30 alphanumerical characters.',
+    );
   }
-  if (!groupName.match(/^[0-9a-zA-Z]+$/)) {
-    flash(
-      'error',
+  if (!group_name.match(/^[0-9a-zA-Z]+$/)) {
+    throw new GroupOperationError(
       'The group name is invalid. Only alphanumerical characters (letters and digits) are allowed.',
     );
-    return;
   }
 
-  // This is technically susceptible to race conditions. That won't be an
-  // issue once we have a unique constraint for group membership.
-  const existingGroupId = await getGroupId(assessmentId, userId);
-  if (existingGroupId != null) {
-    flash('error', 'You are already in a group.');
-    return;
+  if (uids.length === 0) {
+    throw new GroupOperationError('There must be at least one user in the group.');
   }
 
   try {
-    await sqldb.queryAsync(sql.create_group, {
-      assessment_id: assessmentId,
-      user_id: userId,
-      authn_user_id: authnUserId,
-      group_name: groupName,
+    await sqldb.runInTransactionAsync(async () => {
+      let group_id;
+      try {
+        group_id = await sqldb.queryRow(
+          sql.create_group,
+          { assessment_id, authn_user_id, group_name },
+          IdSchema,
+        );
+      } catch (err) {
+        // TODO Check if the error was a duplicate name error
```

Comment:
Do you want to address this in this PR?

Reply:
Thanks for the reminder, done in 706d21c2534720c866bd478917dd96d19a71f610. Are you ok with the approach of checking the error code and constraint?

Reply:
Yes, this looks good to me! Consider adding a link to https://www.postgresql.org/docs/current/errcodes-appendix.html?